compiler/src/iree/compiler/Dialect/Util/Transforms/ConvertPrimitiveType.cpp

```diff
@@ -210,11 +210,108 @@ struct ConvertTypeSensitiveArithCastOp : public OpConversionPattern<OpTy> {
   }
 };
 
+// Truncates input operand if it is a constant and is out of range of target
+// type.
+Value truncateOutOfRangeValue(Value operand,
+                              ConversionPatternRewriter &rewriter,
+                              TypeConverter *typeConverter) {
+  if (!operand.getDefiningOp()) {
+    return operand;
+  }
+  auto oldType = operand.getType();
+  auto newType = typeConverter->convertType(oldType);
+  if (oldType == newType) {
+    return operand;
+  }
+  auto constOp = dyn_cast<arith::ConstantOp>(operand.getDefiningOp());
+  if (!constOp) {
+    return operand;
+  }
+  auto intAttr = constOp->getAttr("value").dyn_cast<IntegerAttr>();
+  APInt value = intAttr.getValue();
+  APInt newValue = value.truncSSat(newType.getIntOrFloatBitWidth());
+  if (newValue == value) {
+    return operand;
+  }
+  Value truncatedConst = rewriter.create<arith::ConstantOp>(
+      constOp->getLoc(), IntegerAttr::get(oldType, newValue));
```

Comment:
Why this is still using `oldType` here?

Reply:
Done! this was legacy code, since it used to be a different pass, it just demoted the value and wait for ConvertType to actually demote the type. Also it used to have issue with other operand type not being demoted yet causing error.